internal/builtin/providers/tf/provider_functions.go

```diff
@@ -0,0 +1,194 @@
+	//Prepend the main error
+	errs := append([]error{m}, diag.Errs()...)
+	return errors.Join(errs...)
+}
+
+func (f *decodeTFVarsFunc) Call(args []cty.Value) (cty.Value, error) {
+	varsFileContent := args[0].AsString()
+	schema, diag := hclsyntax.ParseConfig([]byte(varsFileContent), "", hcl.Pos{Line: 0, Column: 0})
+	if schema == nil || diag.HasErrors() {
+		return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+	}
+	attrs, diag := schema.Body.JustAttributes()
+	// Check if there are any errors.
+	// attrs == nil does not mean that there are no attributes, attrs - is still initialized as an empty map
+	if attrs == nil || diag.HasErrors() {
+		return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+	}
+	vals := make(map[string]cty.Value)
+	for name, attr := range attrs {
+		val, diag := attr.Expr.Value(nil)
+		if diag.HasErrors() {
+			return cty.NullVal(cty.DynamicPseudoType), wrapDiagErrors(FailedToDecodeError, diag)
+		}
+		vals[name] = val
+	}
+	return cty.ObjectVal(vals), nil
+}
+
+// encodeTFVarsFunc encodes an object into a string with the same format as a TFVars file
+type encodeTFVarsFunc struct{}
+
+func (f *encodeTFVarsFunc) Name() string {
+	return "encode_tfvars"
+}
+
+func (f *encodeTFVarsFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name: "input",
+			// The input type is determined at runtime
+			Type:              cty.DynamicPseudoType,
+			Description:       "input to encode for TFVars file. Must be an object with key that are valid identifiers",
+			DescriptionFormat: providers.TextFormattingPlain,
+		},
+	}
+	return providers.FunctionSpec{
+		Parameters:        params,
+		Return:            cty.String,
+		Summary:           "encode an object into a string with the same format as a TFVars file",
+		Description:       "provider::terraform::encode_tfvars encodes an object into a string with the same format as a TFVars file",
+		DescriptionFormat: providers.TextFormattingPlain,
+	}
+}
+
+var InvalidInputError = errors.New("invalid input")
+
+func (f *encodeTFVarsFunc) Call(args []cty.Value) (cty.Value, error) {
+	toEncode := args[0]
+	// null is invalid input
+	if toEncode.IsNull() {
+		return cty.NullVal(cty.String), fmt.Errorf("%w: must not be null", InvalidInputError)
+	}
+	if !toEncode.Type().IsObjectType() {
+		return cty.NullVal(cty.String), fmt.Errorf("%w: must be an object", InvalidInputError)
+	}
+	ef := hclwrite.NewEmptyFile()
+	body := ef.Body()
+
+	// Iterate over the elements of the input value
+	it := toEncode.ElementIterator()
+	for it.Next() {
+		key, val := it.Element()
+		// Check if the key is a string, known and not null, otherwise AsString method panics
+		if !key.Type().Equals(cty.String) || !key.IsKnown() || key.IsNull() {
+			return cty.NullVal(cty.String), fmt.Errorf("%w: object key must be a string: %v", InvalidInputError, key) //TODO errors
```

Comment:
Is this TODO relevant?

Reply:
Not anymore no, thanks for catching that!